File: model/PixivTags.py
# -*- coding: utf-8 -*-
# pylint: disable=I0011, C, C0302

import json
import os
import re
from typing import Optional

import common.PixivHelper as PixivHelper
from common.PixivException import PixivException


class PixivTagsItem:
    imageId: int = 0
    bookmarkCount: int = 0
    imageResponse: int = 0
    ai_type: int = -1

    def __init__(self, image_id, bookmark_count, image_response_count, ai_type=-1):
        self.imageId = image_id
        self.bookmarkCount = bookmark_count
        self.imageResponse = image_response_count
        self.ai_type = ai_type


class PixivTags:
    '''Class for parsing tags search page'''
    itemList = None
    haveImage = None
    isLastPage = None
    availableImages = 0
    # __re_illust = re.compile(r'member_illust.*illust_id=(\d*)')
    # __re_imageItemClass = re.compile(r".*\bimage-item\b.*")
    query = ""
    memberId = 0

    POSTS_PER_PAGE = 60
    page = -1

    def parseMemberTags(self, artist, memberId, query=""):
        '''process artist result and return the image list, https://www.pixiv.net/ajax/user/25661139/illustmanga/tag/<search_tags>'''
        self.itemList = list()
        self.memberId = memberId
        self.query = query
        self.haveImage = artist.haveImages
        self.isLastPage = artist.isLastPage
        for image in artist.imageList:
            self.itemList.append(PixivTagsItem(int(image), 0, 0))

    def parseTags(self, page, query="", curr_page=1):
        '''From search by tags page, https://www.pixiv.net/ajax/search/artworks/<search_tags>'''
        payload = json.loads(page)
        self.query = query
        self.page = curr_page

        # check error
        if payload["error"]:
            raise PixivException(f'Image Error: {payload["message"]}', errorCode=PixivException.SERVER_ERROR)

        # parse images information
        self.itemList = list()
        ad_container_count = 0
        for item in payload["body"]["illustManga"]["data"]:
            if "isAdContainer" in item and item["isAdContainer"]:
                ad_container_count = ad_container_count + 1
                continue

            image_id = int(item["id"])
            # like count not available anymore, need to call separate request...
            bookmarkCount = 0
            imageResponse = 0
            ai_type = -1
            if "aiType" in item:
                ai_type = int(item["aiType"])
            tag_item = PixivTagsItem(image_id, bookmarkCount, imageResponse, ai_type)
            self.itemList.append(tag_item)

        self.haveImage = False
        if len(self.itemList) > 0:
            self.haveImage = True

        # search page info
        self.availableImages = int(payload["body"]["illustManga"]["total"])
        # assume it always return 60 images, including the advert
        if len(self.itemList) + ad_container_count == PixivTags.POSTS_PER_PAGE:
            self.isLastPage = False
        else:
            self.isLastPage = True
        return self.itemList

    def PrintInfo(self):
        PixivHelper.safePrint('Search Result')
        if self.memberId > 0:
            PixivHelper.safePrint(f'Member Id: {self.memberId}')
        PixivHelper.safePrint(f'Query: {self.query}')
        PixivHelper.safePrint(f'haveImage  : {self.haveImage}')
        PixivHelper.safePrint(f'urls  : {len(self.itemList)}')
        for item in self.itemList:
            print(f"\tImage Id: {item.imageId}\tFav Count:{item.bookmarkCount}")
        PixivHelper.safePrint(f'total : {self.availableImages}')
        PixivHelper.safePrint(f'last? : {self.isLastPage}')

    @staticmethod
    def parseTagsList(filename):
        '''read tags.txt and return the tags list'''
        tags = list()

        if not os.path.exists(filename):
            raise PixivException(f"File doesn't exists or no permission to read: {filename}", PixivException.FILE_NOT_EXISTS_OR_NO_PERMISSION)

        reader = PixivHelper.open_text_file(filename)
        for line in reader:
            if line.startswith('#') or len(line) < 1:
                continue
            line = line.strip()
            if len(line) > 0:
                tags.append(line)
        reader.close()
        return tags

class PixivTag:
    """
    AJAX response of https://www.pixiv.net/ajax/search/tags/{tag_id}
    """

    class PixPediaInfo:
        abstract: str
        image: str
        id: str
        tag: str

        def __init__(self, payload):
            self.abstract = payload.get("abstract", "")
            self.image = payload.get("image", "")
            self.id = payload.get("id", "")
            self.tag = payload.get("tag", "")

    tag: str
    word: str
    pixpedia: Optional[PixPediaInfo]
    myFavoriteTags: list[str]
    tagTranslation: dict[str, dict[str, dict]]  # tag_id -> { lang -> translation }

    def __init__(self, payload):
        if payload is None:
            raise PixivException("Tag payload is empty", errorCode=PixivException.OTHER_ERROR)
        if payload.get("error"):
            raise PixivException(payload.get("message", "Tag info error"),
                                 errorCode=PixivException.OTHER_ERROR,
                                 htmlPage=payload)

        body = payload.get("body", {})
        self.tag = body.get("tag", "")
        self.word = body.get("word", "")
        pixpedia_payload = body.get("pixpedia")
        self.pixpedia = self.PixPediaInfo(pixpedia_payload) if pixpedia_payload else None
        self.myFavoriteTags = body.get("myFavoriteTags", [])
        self.tagTranslation = body.get("tagTranslation", {})

File: lib/model/pixiv_tags.dart
/// PixivTags model: tag search results, tag list parsing, and tag info.
library;

import 'dart:convert';
import 'dart:io';

import '../common/pixiv_exception.dart';
import '../common/pixiv_helper.dart' as pixiv_helper;
import 'pixiv_artist.dart';

class PixivTagsItem {
  final int imageId;
  final int bookmarkCount;
  final int imageResponse;
  final int aiType;

  PixivTagsItem(this.imageId, this.bookmarkCount, this.imageResponse,
      [this.aiType = -1]);
}

class PixivTags {
  static const int POSTS_PER_PAGE = 60;

  List<PixivTagsItem>? itemList;
  bool? haveImage;
  bool? isLastPage;
  int availableImages = 0;
  String query = '';
  int memberId = 0;
  int page = -1;

  /// Process member result and return the image list.
  void parseMemberTags(PixivArtist artist, int memberId,
      {String query = ''}) {
    itemList = [];
    this.memberId = memberId;
    this.query = query;
    haveImage = artist.haveImages;
    isLastPage = artist.isLastPage;
    for (final image in artist.imageList) {
      itemList!.add(PixivTagsItem(image, 0, 0));
    }
  }

  /// Parse from search-by-tags page.
  List<PixivTagsItem> parseTags(String page,
      {String query = '', int currPage = 1}) {
    final payload = jsonDecode(page) as Map<String, dynamic>;
    this.query = query;
    this.page = currPage;
    if (payload['error'] == true) {
      throw PixivException(
        'Image Error: ${payload['message']}',
        errorCode: PixivException.SERVER_ERROR,
      );
    }
    itemList = [];
    int adContainerCount = 0;
    for (final item in payload['body']['illustManga']['data'] as List) {
      if (item is Map && item['isAdContainer'] == true) {
        adContainerCount++;
        continue;
      }
      final imageId = int.parse('${item['id']}');
      var aiType = -1;
      if (item['aiType'] != null) aiType = (item['aiType'] as num).toInt();
      itemList!.add(PixivTagsItem(imageId, 0, 0, aiType));
    }

    haveImage = itemList!.isNotEmpty;
    availableImages =
        (payload['body']['illustManga']['total'] as num).toInt();
    isLastPage =
        itemList!.length + adContainerCount != PixivTags.POSTS_PER_PAGE;
    return itemList!;
  }

  void printInfo() {
    pixiv_helper.safePrint('Search Result');
    if (memberId > 0) {
      pixiv_helper.safePrint('Member Id: $memberId');
    }
    pixiv_helper.safePrint('Query: $query');
    pixiv_helper.safePrint('haveImage  : $haveImage');
    pixiv_helper.safePrint('urls  : ${itemList?.length ?? 0}');
    if (itemList != null) {
      for (final item in itemList!) {
        print(
            '\tImage Id: ${item.imageId}\tFav Count:${item.bookmarkCount}');
      }
    }
    pixiv_helper.safePrint('total : $availableImages');
    pixiv_helper.safePrint('last? : $isLastPage');
  }

  /// Read tags.txt and return the list.
  static Future<List<String>> parseTagsList(String filename) async {
    if (!await File(filename).exists()) {
      throw PixivException(
        "File doesn't exists or no permission to read: $filename",
        errorCode: PixivException.FILE_NOT_EXISTS_OR_NO_PERMISSION,
      );
    }
    final lines = await pixiv_helper.openTextFileLines(filename);
    final tags = <String>[];
    for (final line in lines) {
      if (line.startsWith('#') || line.isEmpty) continue;
      final t = line.trim();
      if (t.isNotEmpty) tags.add(t);
    }
    return tags;
  }
}

class PixPediaInfo {
  final String abstract;
  final String image;
  final String id;
  final String tag;

  PixPediaInfo(Map<String, dynamic> payload)
      : abstract = (payload['abstract'] ?? '').toString(),
        image = (payload['image'] ?? '').toString(),
        id = (payload['id'] ?? '').toString(),
        tag = (payload['tag'] ?? '').toString();
}

/// AJAX response of `https://www.pixiv.net/ajax/search/tags/{tag_id}`.
class PixivTag {
  late String tag;
  late String word;
  PixPediaInfo? pixpedia;
  late List<String> myFavoriteTags;
  late Map<String, dynamic> tagTranslation;

  PixivTag(Map<String, dynamic>? payload) {
    if (payload == null) {
      throw PixivException('Tag payload is empty',
          errorCode: PixivException.OTHER_ERROR);
    }
    if (payload['error'] == true) {
      throw PixivException(
        (payload['message'] ?? 'Tag info error').toString(),
        errorCode: PixivException.OTHER_ERROR,
        htmlPage: payload,
      );
    }
    final body = payload['body'] as Map<String, dynamic>? ?? const {};
    tag = (body['tag'] ?? '').toString();
    word = (body['word'] ?? '').toString();
    final p = body['pixpedia'];
    pixpedia = (p is Map<String, dynamic>) ? PixPediaInfo(p) : null;
    myFavoriteTags =
        (body['myFavoriteTags'] as List? ?? const []).map((e) => '$e').toList();
    tagTranslation =
        (body['tagTranslation'] as Map<String, dynamic>? ?? {});
  }
}
